File: sdk/python/dgc_sdk/usage.py
"""Per-run usage records and cost. Prices are embedder-supplied; DGC does not invent tariffs.

Token counts come from the provider usage DGC reports for the session (the ``context`` event's
``input_tokens`` / ``output_tokens`` / ``cached_input_tokens`` totals); a run records the change
across its turns. When the provider reported nothing, the count is ``None`` (unknown), never 0.
``token_estimate`` is DGC's estimate of the conversation's context size, not billed tokens.
"""
from __future__ import annotations

import json
import threading
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Mapping

USAGE_TOTAL_KEYS = ("input_tokens", "output_tokens", "cached_input_tokens", "reasoning_tokens",
                    "requests")


@dataclass(frozen=True)
class Pricing:
    """USD per 1,000,000 tokens. Both sides optional; missing side counts as 0.

    ``cached_input_per_million`` prices the part of the input the provider served from its
    cache; left at 0 it is billed at ``input_per_million``.
    """

    input_per_million: float = 0.0
    output_per_million: float = 0.0
    cached_input_per_million: float = 0.0


def cost_usd(input_tokens: int | None, output_tokens: int | None, cached_input_tokens: int | None,
             pricing: Pricing | None) -> float | None:
    """Cost of one usage record, or None without pricing or when the token counts are unknown.

    ``input_tokens`` counts every prompt token, cached ones included (as DGC reports them);
    ``cached_input_tokens`` is the part of it served from a cache, billed at the cached price.
    """
    if pricing is None or input_tokens is None or output_tokens is None:
        return None
    total_in = max(0, input_tokens)
    cached = min(total_in, max(0, cached_input_tokens or 0))
    cached_rate = float(pricing.cached_input_per_million) or float(pricing.input_per_million)
    dollars = (
        ((total_in - cached) / 1_000_000) * float(pricing.input_per_million)
        + (cached / 1_000_000) * cached_rate
        + (max(0, output_tokens) / 1_000_000) * float(pricing.output_per_million)
    )
    return round(dollars, 8)


def _count(value: Any) -> int | None:
    if isinstance(value, bool):
        return None
    try:
        number = int(value)
    except (TypeError, ValueError):
        return None
    return number if number >= 0 else None


def tokens_from_usage_map(usage: Mapping[str, Any]) -> tuple[int, int, int, int]:
    """Return input, output, cached, estimate; a missing count reads as 0 (see usage_totals)."""
    def _n(*keys: str) -> int:
        for key in keys:
            number = _count(usage.get(key))
            if number is not None:
                return number
        return 0
    inp = _n("input_tokens", "prompt_tokens")
    out = _n("output_tokens", "completion_tokens")
    cached = _n("cached_input_tokens", "cache_read_input_tokens")
    estimate = _n("token_estimate")
    return inp, out, cached, estimate


def usage_totals(event: Mapping[str, Any]) -> dict[str, int] | None:
    """The cumulative session totals a ``context`` event carries, or None if it has none."""
    totals: dict[str, int] = {}
    for key in USAGE_TOTAL_KEYS:
        number = _count(event.get(key))
        if number is None:
            if key in ("input_tokens", "output_tokens", "requests"):
                return None
            number = 0
        totals[key] = number
    return totals


def usage_delta(before: Mapping[str, int], after: Mapping[str, int]) -> dict[str, int] | None:
    """What one turn added, or None when the totals were reset (a new or resumed session)."""
    delta = {key: int(after.get(key, 0)) - int(before.get(key, 0)) for key in USAGE_TOTAL_KEYS}
    if any(value < 0 for value in delta.values()):
        return None
    return delta


def reported(delta: Mapping[str, int]) -> bool:
    """False when requests were made but the provider reported no token usage for them."""
    if int(delta.get("requests", 0)) <= 0:
        return True
    return bool(int(delta.get("input_tokens", 0)) or int(delta.get("output_tokens", 0)))


class UsageLog:
    """Append-only JSONL under the isolated state_dir. Queryable without dgc serve."""

    def __init__(self, directory: Path):
        from ._state import private_dir
        self.directory = private_dir(Path(directory))
        self._lock = threading.Lock()
        self.path = self.directory / "usage.jsonl"

    def record(self, row: Mapping[str, Any]) -> None:
        from .audit import _open_private_append
        payload = dict(row)
        payload.setdefault("ts", time.time())
        line = json.dumps(payload, ensure_ascii=False, default=str) + "\n"
        with self._lock:
            with _open_private_append(self.path) as handle:
                handle.write(line)

    def query(self, *, department: str | None = None, since: float | None = None,
              until: float | None = None) -> dict[str, Any]:
        """Totals over recorded runs. ``unknown_usage_runs`` counts runs whose provider did not
        report tokens; their tokens and cost are excluded from the sums, not counted as 0."""
        rows = []
        if self.path.is_file():
            with self.path.open(encoding="utf-8") as handle:
                for line in handle:
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        row = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    if not isinstance(row, dict):
                        continue
                    if department and str(row.get("department") or "") != department:
                        continue
                    ts = float(row.get("ts") or 0)
                    if since is not None and ts < since:
                        continue
                    if until is not None and ts > until:
                        continue
                    rows.append(row)
        known = [row for row in rows if _row_known(row)]
        costs = [float(row["cost_usd"]) for row in known
                 if isinstance(row.get("cost_usd"), (int, float))]
        return {
            "runs": len(rows),
            "unknown_usage_runs": len(rows) - len(known),
            "input_tokens": sum(_count(row.get("input_tokens")) or 0 for row in known),
            "output_tokens": sum(_count(row.get("output_tokens")) or 0 for row in known),
            "cached_input_tokens": sum(_count(row.get("cached_input_tokens")) or 0 for row in known),
            "cost_usd": round(sum(costs), 8) if costs else None,
            "by_department": _by_department(rows),
            "rows": rows[-500:],
        }


def _row_known(row: Mapping[str, Any]) -> bool:
    return _count(row.get("input_tokens")) is not None and _count(row.get("output_tokens")) is not None


def _by_department(rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    buckets: dict[str, dict[str, Any]] = {}
    for row in rows:
        key = str(row.get("department") or "")
        slot = buckets.setdefault(key, {
            "department": key, "runs": 0, "unknown_usage_runs": 0, "input_tokens": 0,
            "output_tokens": 0, "cost_usd": 0.0,
        })
        slot["runs"] += 1
        if not _row_known(row):
            slot["unknown_usage_runs"] += 1
            continue
        slot["input_tokens"] += _count(row.get("input_tokens")) or 0
        slot["output_tokens"] += _count(row.get("output_tokens")) or 0
        if isinstance(row.get("cost_usd"), (int, float)):
            slot["cost_usd"] += float(row["cost_usd"])
    return list(buckets.values())

File: sdk/typescript/src/usage.ts
/**
 * Per-run usage records and cost. Prices are embedder-supplied; DGC does not invent tariffs.
 * Mirrors sdk/python/dgc_sdk/usage.py.
 *
 * Token counts come from the provider usage DGC reports for the session (the `context` event's
 * cumulative `input_tokens` / `output_tokens` / `cached_input_tokens` totals); a run records the
 * change across its turns. When the provider reported nothing, the count is `null` (unknown),
 * never 0. `token_estimate` is DGC's estimate of the conversation's context size, not billed
 * tokens.
 */
import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { appendPrivate, privateDir } from "./state.ts";
import type { Pricing } from "./types.ts";

export const USAGE_TOTAL_KEYS = [
  "input_tokens", "output_tokens", "cached_input_tokens", "reasoning_tokens", "requests",
] as const;
export type UsageKey = typeof USAGE_TOTAL_KEYS[number];
export type UsageTotals = Record<UsageKey, number>;

export function emptyTotals(): UsageTotals {
  return { input_tokens: 0, output_tokens: 0, cached_input_tokens: 0, reasoning_tokens: 0, requests: 0 };
}

export function count(value: unknown): number | null {
  if (typeof value === "boolean" || value === null || value === undefined || value === "") return null;
  const number = Number(value);
  if (!Number.isFinite(number)) return null;
  const whole = Math.trunc(number);
  return whole >= 0 ? whole : null;
}

/**
 * Cost of one usage record, or null without pricing or when the token counts are unknown.
 * `inputTokens` counts every prompt token, cached ones included (as DGC reports them);
 * `cachedInputTokens` is the part served from a cache, billed at the cached price.
 */
export function costUsd(inputTokens: number | null, outputTokens: number | null,
  cachedInputTokens: number | null, pricing: Pricing | undefined): number | null {
  if (!pricing || inputTokens === null || outputTokens === null) return null;
  const totalIn = Math.max(0, inputTokens);
  const cached = Math.min(totalIn, Math.max(0, cachedInputTokens || 0));
  const inputRate = Number(pricing.inputPerMillion || 0);
  const cachedRate = Number(pricing.cachedInputPerMillion || 0) || inputRate;
  const dollars = ((totalIn - cached) / 1_000_000) * inputRate
    + (cached / 1_000_000) * cachedRate
    + (Math.max(0, outputTokens) / 1_000_000) * Number(pricing.outputPerMillion || 0);
  return Math.round(dollars * 1e8) / 1e8;
}

/** The cumulative session totals a `context` event carries, or null if it has none. */
export function usageTotals(event: Record<string, unknown>): UsageTotals | null {
  const totals = emptyTotals();
  for (const key of USAGE_TOTAL_KEYS) {
    const number = count(event[key]);
    if (number === null) {
      if (key === "input_tokens" || key === "output_tokens" || key === "requests") return null;
      continue;
    }
    totals[key] = number;
  }
  return totals;
}

/** What one turn added, or null when the totals were reset (a new or resumed session). */
export function usageDelta(before: UsageTotals, after: UsageTotals): UsageTotals | null {
  const delta = emptyTotals();
  for (const key of USAGE_TOTAL_KEYS) {
    delta[key] = (after[key] || 0) - (before[key] || 0);
    if (delta[key] < 0) return null;
  }
  return delta;
}

/** False when requests were made but the provider reported no token usage for them. */
export function reported(delta: UsageTotals): boolean {
  if ((delta.requests || 0) <= 0) return true;
  return Boolean(delta.input_tokens || delta.output_tokens);
}

export type UsageReport = {
  runs: number;
  /** Runs whose provider did not report tokens; their tokens and cost are left out of the sums. */
  unknownUsageRuns: number;
  inputTokens: number;
  outputTokens: number;
  cachedInputTokens: number;
  /** Sum over runs with a known cost, or null when none has one. */
  costUsd: number | null;
  byDepartment: Array<{
    department: string; runs: number; unknownUsageRuns: number; inputTokens: number;
    outputTokens: number; costUsd: number;
  }>;
  /** The last 500 rows as recorded (snake_case keys, the same JSONL the Python SDK writes). */
  rows: Array<Record<string, unknown>>;
};

function rowKnown(row: Record<string, unknown>): boolean {
  return count(row.input_tokens) !== null && count(row.output_tokens) !== null;
}

/** Append-only JSONL under the isolated stateDir. Queryable without dgc serve. */
export class UsageLog {
  readonly directory: string;
  readonly path: string;

  constructor(directory: string) {
    this.directory = privateDir(directory);
    this.path = join(this.directory, "usage.jsonl");
  }

  record(row: Record<string, unknown>): void {
    const payload = { ts: Date.now() / 1000, ...row };
    appendPrivate(this.path, JSON.stringify(payload) + "\n");
  }

  query(options: { department?: string; since?: number; until?: number } = {}): UsageReport {
    const rows: Array<Record<string, unknown>> = [];
    if (existsSync(this.path)) {
      for (const line of readFileSync(this.path, "utf8").split("\n")) {
        if (!line.trim()) continue;
        let row: unknown;
        try { row = JSON.parse(line); } catch { continue; }
        if (!row || typeof row !== "object" || Array.isArray(row)) continue;
        const record = row as Record<string, unknown>;
        if (options.department && String(record.department || "") !== options.department) continue;
        const ts = Number(record.ts || 0);
        if (options.since !== undefined && ts < options.since) continue;
        if (options.until !== undefined && ts > options.until) continue;
        rows.push(record);
      }
    }
    const known = rows.filter(rowKnown);
    const costs = known.map((row) => row.cost_usd).filter((value): value is number => typeof value === "number");
    const buckets = new Map<string, UsageReport["byDepartment"][number]>();
    for (const row of rows) {
      const key = String(row.department || "");
      let slot = buckets.get(key);
      if (!slot) {
        slot = { department: key, runs: 0, unknownUsageRuns: 0, inputTokens: 0, outputTokens: 0, costUsd: 0 };
        buckets.set(key, slot);
      }
      slot.runs += 1;
      if (!rowKnown(row)) {
        slot.unknownUsageRuns += 1;
        continue;
      }
      slot.inputTokens += count(row.input_tokens) || 0;
      slot.outputTokens += count(row.output_tokens) || 0;
      if (typeof row.cost_usd === "number") slot.costUsd += row.cost_usd;
    }
    const sum = (key: string) => known.reduce((total, row) => total + (count(row[key]) || 0), 0);
    return {
      runs: rows.length,
      unknownUsageRuns: rows.length - known.length,
      inputTokens: sum("input_tokens"),
      outputTokens: sum("output_tokens"),
      cachedInputTokens: sum("cached_input_tokens"),
      costUsd: costs.length ? Math.round(costs.reduce((a, b) => a + b, 0) * 1e8) / 1e8 : null,
      byDepartment: [...buckets.values()],
      rows: rows.slice(-500),
    };
  }
}
